### ISSUE TITLE
[Feature Enhancement]  Multi-build capable asset resources

### ISSUE THREAD

**Issue by jspaleta** [sensu/sensu-go#3067]
Currently  named assets are effectively tied to a single type of agent host, leading to highly duplicative configuration in heterogeneous environments.

Asset filter works to prevent the wrong binary from being installed on an agent, but there's no mechanism by which an agent to choose among multiple binaries.

Examples: 
1) you would like to run a statically compiled golang check on both osX and linux.
  currently you'd have to have a different named asset and checks for each target OS, in order for the asset download to work correctly.
2) you would like to use Sensu ruby-runtime check on both Alpine Linux and Centos Linux.
Both would require different named assets, and checks for each OS.

## Expected Behavior
Have all possible asset builds in a single Asset resource, so named asset resources can be re-used in check definitions for agents to use regardless of agent's host OS/platform/arch
Ex: 
asset named "sensu-plugins/sensu-plugins-cpu-checks" should list all asset binaries+filters, and should be usable across all checks and subscriptions. The agent should be able to determine which binary to use based on the filters.

## Current Behavior
Difficult to impossible to re-use a named asset resource in a heterogeneous environment.  The result is effectively building up check,asset,subscription naming schemes for different platforms.

## Strawman:
introduce builds list  in Asset spec
each build is a hash of key/value pairs:
```
  url: (required) the location of the binary
  sha512: (required) the sha512sum
  priority: (optional) numeric used by asset manager to help choose if multiple build pass filter test. should be unique for each build, defaults to 0
  name:  (optional) human readable name for the build, defaults to null
  filters: (optional) the list of entity filters that must match, defaults to null
```
See Gist: https://gist.github.com/jspaleta/01c9fa0eb2b22ace25ae03c01e47b368

###  Notes:
* To achieve deterministic behavior when multiple builds pass filter matching, priority should be unique for each build. highest priority build with all filter conditions satisfied wins. 

* This can be implemented as an extension to existing v2 asset api, if in addition the high level spec attributes "name" and "priority" are added as well. Asset agent could be taught to fold the build outside the build list into its build list representation as an additional member.

* Could be implemented as v3 api and drop the high level sha512, url, filters keys and require all information to exist in the builds list.

### Key benefits
* Extends Asset resource in a backward compatible way, so existing asset configs without build list continue to work.
* Allow to encode a single named asset configuration in a namespace for all asset binaries
* Simplifies Bonsai UI, no need for platform/arch modular dialog.
* Potentially simplifies future bonsai asset lifecycle tooling -> mapping single asset resource to each bonsai index listing.

**Comment by calebhailey** [comment]
I was mocking some ideas around this a while back too; here's what I came up with: https://gist.github.com/calebhailey/00d964cc50fa63af3ef1be50adc30376 

**TL;DR** add a `builds` scope to a new v3 Asset spec. Seems like you could do this in a backwards compatible way (e.g. if `spec.url` and `spec.sha512` are defined, make them build 0 (in the array/slice of builds). 

```yaml
---
type: Asset
api_version: core/v3
metadata:
  name: monitoring-plugins
  namespace: default
spec:
  build:
  - url: https://github.com/calebhailey/monitoring-plugins-assets/releases/download/v2.2-beta/alpine-monitoring-plugins-v2.2-linux-x86_64.tar.gz
    sha512: 5b68a4b48168e4af56a830c478bb281b333103d908bc3db5dd945d891bba8ad40f58203904e2614dab55e362ec6ef7737ba3e3efd02f4703d08eb1b66dc03983
    platform: alpine
    os: linux
```

**Comment by calebhailey** [comment]
Huge understatement: 

> Simplifies Bonsai UI, no need for platform/arch modular dialog. 

Just find an asset & download! :boom:

**Comment by jspaleta** [comment]
Note on the GOARM prereq.
I agree that sorting out the GOARM stuff will be imporant before we enable multi-build assets in bonsai.

But we can still test this as an undocumented feature or experimental feature ahead of GOARM
Users always have the option to write a filter that matches an entity annotation instead of system attribute if they need to flag their armv6 versus armv7 compatible agents.
